Fix thin hyper shell colorization of top boundary

gassmoeller opened the pull request:
The function GridGenerator::quarter_hyper_shell would colorize the boundary ids of the triangulation incorrectly if the created shell was too thin. In particular the top and bottom boundary ids would not be set. This did not matter for the bottom boundary since it is 0, which is also the default value, but the top boundary would not be marked correctly for these geometries.

The fix is to not use the center of the face to compute the radius to check for closeness to outer / inner radius, but instead use the average radius of all the vertices of the face. The center (=midpoint in euclidean geometries) has a different radius than all the vertices. All the other line changes are just a renaming a variable that was used all over this function.

The added test fails on master, but passes on this branch.

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -1461,67 +1461,69 @@ namespace GridGenerator
         for (const unsigned int f : GeometryInfo<3>::face_indices())
           {
-            if (!cell->face(f)->at_boundary())
+            const auto &face = cell->face(f);
```
`cell->face(f)` returns an object, not a reference to something:
```suggestion
            const auto face = cell->face(f);
```

bangerth commented on source/grid/grid_generator.cc:
```diff
@@ -1461,67 +1461,69 @@ namespace GridGenerator
+            // radius of all vertices instead.
+            double radius = 0.0;
+            for (unsigned int v : face->vertex_indices())
+              radius += face->vertex(v).norm() - center.norm();
```
That's not what you want. You want this instead:
```suggestion
              radius += (face->vertex(v) - center).norm();
```

bangerth replied:
I recognize that this was wrong before already.

bangerth reviewed (commented):
Yes, nice. Thanks for tracking this down! Since I'm already asking you to make a couple of changes to the file, would you mind also making the two variables in line 1457, 1458 `const` and moving the `cell` variable in line 1459 into the loop header?

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1454,74 +1454,75 @@ namespace GridGenerator
               {
-                cell->face(f)->set_boundary_id(2);
+                face->set_boundary_id(2);
                 for (unsigned int j = 0; j < GeometryInfo<3>::lines_per_face;
                      ++j)
```
```suggestion
                for (const unsigned int line_no : face->line_indices())
```
if we are using the range-for above then we can improve consistency by changing this too.

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -1454,74 +1454,75 @@ namespace GridGenerator
-            double radius = cell->face(f)->center().norm() - center.norm();
-            if (std::fabs(cell->face(f)->center()[0]) <
-                eps) // x = 0 set boundary 2
+            // One could use the radius of the center of the face, but for
+            // curved cells the radius of the (euclidean) center is not
+            // equal to the radius at the vertices. Compute the average
+            // radius of all vertices instead.
```
This is true, and this patch would fix this function, but we should fix it properly by addressing the historical wart which causes this weirdness.

We presently have

```cpp
    if (colorize)
      colorize_quarter_hyper_shell(tria, center, inner_radius, outer_radius);

    tria.set_all_manifold_ids(0);
    tria.set_manifold(0, SphericalManifold<3>(center));
```
because that code was written for the old `Boundary` system (pre-manifolds we needed to use boundary ids to describe geometry). These systems are completely independent now - we might as well just set the manifolds first and then `face->center(true)` will give the correct answer. Would you mind fixing the orders and then using the correct centers in this function? We might need to update more than one place in `grid_generator.cc`.

gassmoeller commented:
Thanks for the comments. I addressed them for the quarter hyper shell. I think the half hyper shell needs similar fixes for the colorization, but I will address that in a follow-up PR to not delay this one.

gassmoeller commented:
/rebuild

gassmoeller commented:
I included the things I learned in #18574 into this PR, i.e. I removed the use of `GeometryInfo` and added output to the test that also records the boundary ids of lines in 3D (for future reference).

This is now ready from my side.
